pkg/cmd/util/clientcmd/factory.go

```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 	c.clients[config.GroupVersion.String()] = client
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(f *Factory) ([]unversioned.GroupResource, error) {
+	m, _ := f.Object(false)
+	oclient, _, err := f.Clients()
+	if err != nil {
+		return nil, err
+	}
+	set := make(map[string]unversioned.GroupResource)
```

Comment:
At first I didn't want to deal with group resources, but I guess I'm going to have to.